### PR TITLE
(video_shader_parse) Optimise parsing of (non-slang) shader parameters

### PR DESCRIPTION
**Pull request by jdgleaver**
## Description

At present, when parsing non-slang shader parameters (i.e. `#pragma parameter` entries), the shader file is opened and read line-by-line. This kind of file access has proven to be slow on some platforms, so this PR instead reads the shader file in one go, then parses the contents in memory.

This PR also improves the efficiency of detecting `#pragma parameter` entries (we no longer call `sscanf()` on every single line of the file...)

## Reviewers

@twinaphex

### CODE REVIEW

**Comment by twinaphex** [comment]
Tested - seems to work well. Thanks for this!

**Comment by LazyBumHorse** [comment]
Got 2 small question(s):
Why use `path_is_valid()` before `filestream_read_file()`, does the latter not "fail correctly" or is this just faster overall?

https://github.com/libretro/RetroArch/blob/bcfc5db7d8fa02cc45330a5a36c929d0417a9612/gfx/video_shader_parse.c#L522-L523
Why not simply `free(buf);`?
Conversions to `void*` are implicit even in C++ and `free()` does nothing on null pointers - except maybe on pre-standard C compilers (are we using those?)

**Comment by jdgleaver** [comment]
@LazyBumHorse Good questions!

- `filestream_read_file()` does indeed fail gracefully, but a `path_is_valid()` check is ~4 orders of magnitude faster. Note that the real-world performance impact in this case is negligible, but we still like to use `path_is_valid()` whenever it is possible/appropriate to do so.

- Regarding the `(void*)` cast - you got me there! I wanted to get this PR done quickly (it was a request), so I copy/pasted some existing code from `config_file.c`. The cast is a leftover. AFAIK, there is no case in which it is required (but it does no harm either)

- The null pointer check before freeing is more interesting. You'll see that throughout the RetroArch source code, and it's by design. It's a result of RetroArch's goal of complete cross-platform compatibility. You've guessed the reasoning, but let me explain, anyway:  
For all C standard compliant compilers/libraries, `free(NULL)` is indeed a no-op, but compliance is not guaranteed. Imagine that we get rid of all the null checks - looks good, right? The code is cleaner, and we save a few clock cycles. Now imagine that someone ports RetroArch to e.g. Palm OS. Uh oh... now every `free(NULL)` causes a crash...  
So basically, we pre-emptively add null pointer checks to reduce stress levels for future devs who might want to support esoteric platforms.  
(Note: If we're being really pedantic, the check should be an explicit `if (ptr != NULL)`, since on some platforms `NULL != 0`... but this is incredibly obscure, and the affected hardware generally wouldn't have the capacity to run RetroArch anyway, so it's not worth the hassle...)

**Comment by LazyBumHorse** [comment]
@jdgleaver Thanks for the detailed answer!
I'll try to not use `free` on null pointers then.
So far I've always used `malloc` checks á la `if (!ptr) return;` beforehand, which is not great, but it seems we don't have any convention for failing `malloc` - which is fine, for `realloc` on the other hand...

> Note: If we're being really pedantic, the check should be an explicit if (ptr != NULL), since on some platforms NULL != 0...

Hm... this should never be the case, I'll try to explain why (more pedantry incoming).
According to the C standard, any integer constant with value 0 converted to a pointer type is a null pointer and `NULL` is a null pointer of type `void*` (it's practically `(void*)0`), therefore the conditions `(ptr != NULL)`, `(ptr != 0)` and `(ptr)` are all equivalent.
However, this does not mean that the binary value of a null pointer has to be all-bits-0, if you were to type-pun it like
```
void *ptr = NULL;
uint32_t i;
memcpy(&i, &ptr, sizeof i);
```
`i` doesn't have to be 0. This also means that `memset`ing pointers to 0 can fail on some platforms.
Maybe this is what you mean?

I realize that the C standard isn't always followed, but there shouldn't (as in hopefully shouldn't) be any implementation that redefines `NULL` like that.

**Comment by jdgleaver** [comment]
@LazyBumHorse You have a very detailed knowledge of the C standard - thanks for sharing this!

My point with `if (ptr != NULL)` is that, for example, on the C64, a pointer with a value of zero is valid, and is used to access an IO port. So I don't think it would work as you say in this case...

But for everything that RetroArch does/will support, you are of course correct  :)

**Comment by LazyBumHorse** [comment]
@jdgleaver 

> on the C64, a pointer with a value of zero is valid, and is used to access an IO port. So I don't think it would work as you say in this case

Ah, I see what you mean, you'd want to write code like
`*(char *)0 = 1;`
to access the IO port.
If the compiler is standard compliant, this would not work because `(char *)0` is a null pointer which probably has another binary value than 0.
But there is a way out of this, you can type-pun the address - actually, I think there's an easier way, remember how only integer _constants_ convert to null pointers?
You can do this:
```
unsigned adress = 0;
*(char*)adress = 1;
```
there you go, `(char*)adress` will be a pointer of value 0 and not a null pointer.

But yeah, in the end it all comes down to implementation, I can see how not every compiler writer would want to be 100% compliant, especially if it's for their C64 in their basement.

**Comment by jdgleaver** [comment]
@LazyBumHorse Thanks for the extra info! I love how you know these obscure things :)

**Comment by LazyBumHorse** [comment]
@jdgleaver I've studied C quite a bit, you should run before I get into how this is different in C++ :smile:
